### PR TITLE
fix: prevent blacklisted symbols from appearing in `.ilean` files

### LINKED ISSUE #1510: Workspace symbol search returns internal definitions
The workspace symbol search (`ctrl-p #` in vscode) returns all kinds of internal definitions.  Example queries:
 - `_sunfold` returns smart unfolding lemmas
 - `_unsafe_rec` returns helper definitions used to compile recursive definitions

These should be filtered out.

Nightly from August 25.

### PR DESCRIPTION
**Pull request by rish987**
Closes #1510. I'm still trying to figure out how to test this properly, I'll ask about that at the meeting tomorrow. But at least when I run `lean file.lean -i file.ilean` it looks like the undesirable definitions are indeed being filtered out.

### CODE REVIEW

File: src/Lean/Server/Completion.lean [hunk 1]
```diff
@@ -26,14 +26,13 @@ builtin_initialize completionBlackListExt : TagDeclarationExtension ← mkTagDec
...
-  <||> isRec declName
-  <||> (pure <| completionBlackListExt.isTagged env declName)
-  <||> isMatcher declName
+def isBlackListed (env : Environment) (declName : Name) : Bool :=
+  (declName.isInternal && !isPrivateName declName)
+  || isAuxRecursor env declName
+  || isNoConfusion env declName
+  || isRecCore env declName
+  || completionBlackListExt.isTagged env declName
+  || isMatcherCore env declName
```
Comment (rish987) [inline src/Lean/Server/Completion.lean:35]: I was quite unsure about this change. I needed this to be outside of `MetaM` but I'm not certain about the extent to which Lean is now eagerly evaluating the components of this big chained or -- ideally it should stop after the first one that returns true (which I assume was certainly the case in the previous `MetaM` version).

File: src/Lean/Server/References.lean [hunk 2]
```diff
@@ -130,13 +131,18 @@ def identOf : Info → Option (RefIdent × Bool)
   | Info.ofFieldInfo fi => some (RefIdent.const fi.projName, false)
   | _ => none
 
+open Completion in
```
Comment (rish987) [inline src/Lean/Server/References.lean:134]: Having to pull in `Completion` here felt a bit hackish to me, is there a good candidate parent file that I can migrate the `isBlacklisted` related functions to?

File: src/Lean/Server/Completion.lean (hunk 1 above)
Comment (Kha) [reply src/Lean/Server/Completion.lean:35]: That's fine, `||` is lazy

File: src/Lean/Server/References.lean [hunk 3]
```diff
@@ -130,13 +131,18 @@ def identOf : Info → Option (RefIdent × Bool)
...
 def findReferences (text : FileMap) (trees : Array InfoTree) : Array Reference := Id.run <| StateT.run' (s := #[]) do
   for tree in trees do
     tree.visitM' (postNode := fun ci info _ => do
       if let some (ident, isBinder) := identOf info then
-        if let some range := info.range? then
-          if info.stx.getHeadInfo matches .original .. then  -- we are not interested in canonical syntax here
-            modify (·.push { ident, range := range.toLspRange text, stx := info.stx, ci, info, isBinder }))
+        let blacklisted? := match ident with
+        | .const name => isBlackListed ci.env name
+        | _ => false
```
Comment (Kha) [inline src/Lean/Server/References.lean:141]: I would indent these lines

File: src/Lean/Server/References.lean [hunk 4]
```diff
@@ -130,13 +131,18 @@ def identOf : Info → Option (RefIdent × Bool)
...
 
+open Completion in
 def findReferences (text : FileMap) (trees : Array InfoTree) : Array Reference := Id.run <| StateT.run' (s := #[]) do
   for tree in trees do
     tree.visitM' (postNode := fun ci info _ => do
       if let some (ident, isBinder) := identOf info then
-        if let some range := info.range? then
-          if info.stx.getHeadInfo matches .original .. then  -- we are not interested in canonical syntax here
-            modify (·.push { ident, range := range.toLspRange text, stx := info.stx, ci, info, isBinder }))
+        let blacklisted? := match ident with
```
Comment (Kha) [inline src/Lean/Server/References.lean:139]:
```suggestion
        let blacklisted := match ident with
```
It's not an `Option` or `Except`

File: src/Lean/Server/References.lean (hunk 2 above)
Comment (Kha) [reply src/Lean/Server/References.lean:134]: When in doubt, use `Lean.Server.Utils` :)

**Review by Kha — CHANGES_REQUESTED** [review]

**Comment by Kha** [comment]
I'm not sure if this is the best way to solve the issue - it's not clear to me why the `_sunfold`s etc. are in the info tree to begin with, and with an original source span at that.
Note that the issue seems to be limited to `let rec`. @leodemoura Do you see where these info nodes could come from? Perhaps we can investigate in today's meeting.

**Comment by leodemoura** [comment]
In the frontend meeting, I think we all agreed to close this PR. If there are no objections, let's close it.

**Pull request closed without merging**